### PR TITLE
Add Blob Gossip

### PR DESCRIPTION
**Pull request by terencechain**
This PR
- Adds blob p2p validator outlined in: https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#blob_sidecar_index
- Adds sufficient unit tests to cover 81% of `validate_blobs.go`
- Adds blob subscriber, note this mostly is no-op, it simply adds the blobs to the seen cache but still needs to cache the blobs before passing all or nothing to blockchain pkg. (See TODO)
- Similar to block, blob also has a pending cache for unknown parent

### CODE REVIEW

File: beacon-chain/p2p/pubsub_filter.go [hunk 1]
```diff
@@ -57,12 +57,17 @@ func (s *Service) CanSubscribe(topic string) bool {
 		log.WithError(err).Error("Could not determine Capella fork digest")
 		return false
 	}
-
+	denebForkDigest, err := forks.ForkDigestFromEpoch(params.BeaconConfig().DenebForkEpoch, s.genesisValidatorsRoot)
+	if err != nil {
+		log.WithError(err).Error("Could not determine Capella fork digest")
```
Comment (rkapka) [inline beacon-chain/p2p/pubsub_filter.go]:
```suggestion
		log.WithError(err).Error("Could not determine Deneb fork digest")
```

File: beacon-chain/sync/pending_blocks_queue.go [hunk 2]
```diff
@@ -445,6 +460,16 @@ func (s *Service) addPendingBlockToCache(b interfaces.ReadOnlySignedBeaconBlock)
 	return nil
 }
 
+// This adds blob to slotToPendingBlobs cache.
+func (s *Service) addPendingBlobToCache(b *eth.SignedBlobSidecar) error {
+	blobs := s.pendingBlobsInCache(b.Message.Slot)
+
+	blobs = append(blobs, b)
```
Comment (rkapka) [inline beacon-chain/sync/pending_blocks_queue.go:467]: Is it possible that we accidentally append a blob that has been already cached, resulting in having the same blob twice?

File: beacon-chain/sync/service.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ var _ runtime.Service = (*Service)(nil)
 
 const rangeLimit = 1024
 const seenBlockSize = 1000
+const seenBlobSize = 4000 // Each block can have max 4 blobs. Worst case 164kB for cache.
```
Comment (rkapka) [inline beacon-chain/sync/service.go:46]:
```suggestion
`const seenBlobSize = seenBlockSize * 4` // Each block can have max 4 blobs. Worst case 164kB for cache.
```

File: beacon-chain/sync/subscriber.go [hunk 4]
```diff
@@ -130,6 +131,17 @@ func (s *Service) registerSubscribers(epoch primitives.Epoch, digest [4]byte) {
...
 	}
+
+	// New Gossip Topic in Deneb
+	if epoch >= params.BeaconConfig().DenebForkEpoch {
+		s.subscribeStaticWithSubnets(
+			p2p.BlobSubnetTopicFormat,
+			s.validateBlob,   /* validator */
+			s.blobSubscriber, /* message handler */
+			digest,
+			params.BeaconConfig().MaxBlobsPerBlock,
```
Comment (rkapka) [inline beacon-chain/sync/subscriber.go:142]: I think we should have a `BlobsSubnetCount` value in the network config for symmetry with `AttestationSubnetCount`.

File: beacon-chain/sync/subscriber_blob_sidecar.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	"fmt"
+
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"google.golang.org/protobuf/proto"
+)
+
+func (s *Service) blobSubscriber(ctx context.Context, msg proto.Message) error {
+	b, ok := msg.(*eth.SignedBlobSidecar)
+	if !ok {
+		return fmt.Errorf("message was not type *eth.Attestation, type=%T", msg)
```
Comment (rkapka) [inline beacon-chain/sync/subscriber_blob_sidecar.go:14]:
```suggestion
		return fmt.Errorf("message was not type *eth.SignedBlobSidecar, type=%T", msg)
```

File: testing/util/block.go [hunk 6]
```diff
@@ -77,6 +77,19 @@ func NewBeaconBlock() *ethpb.SignedBeaconBlock {
 	}
 }
 
+func NewBlobsidecar() *ethpb.SignedBlobSidecar {
+	return &ethpb.SignedBlobSidecar{
+		Message: &ethpb.BlobSidecar{
+			BlockRoot:       make([]byte, fieldparams.RootLength),
+			BlockParentRoot: make([]byte, fieldparams.RootLength),
+			Blob:            make([]byte, 131072),
```
Comment (rkapka) [inline testing/util/block.go:85]: I would like to see this magic number in `fieldparams`. It's very hard to remember.

File: beacon-chain/sync/validate_blob.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	sBlob, ok := m.(*eth.SignedBlobSidecar)
+	if !ok {
+		log.WithField("message", m).Error("Message is not of type *eth.SignedBlobSidecar")
+		return pubsub.ValidationReject, errWrongMessage
+	}
+	blob := sBlob.Message
+
+	// [REJECT] The sidecar is for the correct topic -- i.e. sidecar.index matches the topic {index}.
+	want := fmt.Sprintf("blob_sidecar_%d", blob.Index)
+	if !strings.Contains(*msg.Topic, want) {
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:49]: Why do we use `Contains` here? Is the topic `123blob_sidecar_0xyz Radek` valid?

File: beacon-chain/sync/validate_blob.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+	genesisTime := uint64(s.cfg.chain.GenesisTime().Unix())
+	if err := slots.VerifyTime(genesisTime, blob.Slot, earlyBlockProcessingTolerance); err != nil {
+		return pubsub.ValidationIgnore, errors.Wrap(err, "blob too far into future")
+	}
+
+	// [IGNORE] The sidecar is from a slot greater than the latest finalized slot --
+	// i.e. validate that sidecar.slot > compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)
+	startSlot, err := slots.EpochStart(s.cfg.chain.FinalizedCheckpt().Epoch)
+	if err != nil {
+		log.WithError(err).WithFields(blobFields(blob)).Error("Ignored block: could not calculate epoch start slot")
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:65]:
```suggestion
		log.WithError(err).WithFields(blobFields(blob)).Error("Ignored blob: could not calculate epoch start slot")
```

File: beacon-chain/sync/validate_blob.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+		return pubsub.ValidationIgnore, err
+	}
+	if startSlot >= blob.Slot {
+		err := fmt.Errorf("finalized slot %d greater or equal to blob slot %d", startSlot, blob.Slot)
+		log.WithFields(blobFields(blob)).Warn(err)
+		return pubsub.ValidationIgnore, err
+	}
+
+	// [IGNORE] The blob's block's parent (defined by sidecar.block_parent_root) has been seen (via both gossip and non-gossip sources)
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:74]:
```suggestion
	// [IGNORE] The sidecars's block's parent (defined by sidecar.block_parent_root) has been seen (via both gossip and non-gossip sources)
```
Nitpick :)

File: beacon-chain/sync/validate_blob.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	parentRoot := bytesutil.ToBytes32(blob.BlockParentRoot)
+	if !s.cfg.chain.HasBlock(ctx, parentRoot) {
+		if err := s.addPendingBlobToCache(sBlob); err != nil {
+			log.WithError(err).WithFields(blobFields(blob)).Error("Failed to add blob to queue")
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:78]:
```suggestion
			log.WithError(err).WithFields(blobFields(blob)).Error("Failed to add blob to cache")
```

File: beacon-chain/sync/validate_blob.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+			return pubsub.ValidationIgnore, err
+		}
+		log.WithFields(blobFields(blob)).Warn("Ignored blob: parent block not found")
+		return pubsub.ValidationIgnore, nil
+	}
+
+	// [REJECT] The sidecar's block's parent (defined by sidecar.block_parent_root) passes validation.
+	blk, err := s.cfg.chain.GetBlock(ctx, parentRoot)
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:86]: Is this enough for "passes validation"?

File: beacon-chain/sync/validate_blob.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+		err := fmt.Errorf("parent block slot %d greater or equal to blob slot %d", blk.Block().Slot(), blob.Slot)
+		log.WithFields(blobFields(blob)).Error(err)
+		return pubsub.ValidationReject, err
+	}
+
+	// [REJECT] The proposer signature, signed_blob_sidecar.signature,
+	// is valid with respect to the sidecar.proposer_index pubkey.
+	parentState, err := s.cfg.stateGen.StateByRoot(ctx, parentRoot)
+	if err != nil {
+		return pubsub.ValidationIgnore, err
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:103]: Shouldn't we reject here? After all we are unable to verify the signature.

File: beacon-chain/sync/validate_blob.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/crypto/bls"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/v4/network/forks"
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	"github.com/sirupsen/logrus"
+)
+
+func (s *Service) validateBlob(ctx context.Context, pid peer.ID, msg *pubsub.Message) (pubsub.ValidationResult, error) {
```
Comment (rkapka) [inline beacon-chain/sync/validate_blob.go:24]: I see a little inconsistency in this function with regards to logging. We usually log before returning on ignore/reject, but not always.

File: beacon-chain/sync/validate_blob.go (hunk 11 above)
Comment (terencechain) [reply beacon-chain/sync/validate_blob.go:86]: It's enough. We only put validated object in the DB or initial sync cache

File: beacon-chain/sync/validate_blob.go (hunk 12 above)
Comment (terencechain) [reply beacon-chain/sync/validate_blob.go:103]: We are not able to get the parent state, which does not mean we cannot verify the signature. In this case, we probably shouldn't penalize the peer, but we do drop the object

File: beacon-chain/sync/validate_blob.go (hunk 10 above)
Comment (terencechain) [reply beacon-chain/sync/validate_blob.go]: dbdf3851f

File: beacon-chain/sync/validate_blob.go (hunk 8 above)
Comment (terencechain) [reply beacon-chain/sync/validate_blob.go]: dbdf3851f

File: beacon-chain/sync/subscriber.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/sync/subscriber.go]: dbdf3851f

File: beacon-chain/sync/pending_blocks_queue.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/pending_blocks_queue.go:475]: Nice catch! dbdf3851f

File: beacon-chain/sync/pending_blocks_queue.go [hunk 14]
```diff
@@ -427,6 +428,20 @@ func (s *Service) pendingBlocksInCache(slot primitives.Slot) []interfaces.ReadOn
...
 
+// This returns signed blob sidecar given input key from slotToPendingBlobs.
+func (s *Service) pendingBlobsInCache(slot primitives.Slot) []*eth.SignedBlobSidecar {
+	k := slotToCacheKey(slot)
+	value, ok := s.slotToPendingBlobs.Get(k)
+	if !ok {
+		return []*eth.SignedBlobSidecar{}
+	}
+	bs, ok := value.([]*eth.SignedBlobSidecar)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/sync/pending_blocks_queue.go:439]: dbdf3851f

File: beacon-chain/blockchain/receive_block.go [hunk 15]
```diff
@@ -26,6 +26,7 @@ type BlockReceiver interface {
 	ReceiveBlock(ctx context.Context, block interfaces.ReadOnlySignedBeaconBlock, blockRoot [32]byte) error
 	ReceiveBlockBatch(ctx context.Context, blocks []interfaces.ReadOnlySignedBeaconBlock, blkRoots [][32]byte) error
 	HasBlock(ctx context.Context, root [32]byte) bool
+	GetResentBlockSlot(root [32]byte) (primitives.Slot, error)
```
Comment (rkapka) [inline beacon-chain/blockchain/receive_block.go:29]: What's the purpose of using the word `Resent` here? Was it supposed to be `Recent`?

File: beacon-chain/sync/pending_blocks_queue.go [hunk 16]
```diff
@@ -445,6 +461,23 @@ func (s *Service) addPendingBlockToCache(b interfaces.ReadOnlySignedBeaconBlock)
 	return nil
 }
 
+// This adds blob to slotToPendingBlobs cache.
+func (s *Service) addPendingBlobToCache(b *eth.SignedBlobSidecar) error {
+	blobs := s.pendingBlobsInCache(b.Message.Slot)
+
+	// If we already have seen the index. Ignore it.
```
Comment (rkapka) [inline beacon-chain/sync/pending_blocks_queue.go:468]:
```suggestion
	// If we already have seen the index, ignore it.
```

File: beacon-chain/blockchain/receive_block.go (hunk 15 above)
Comment (terencechain) [reply beacon-chain/blockchain/receive_block.go]: Renamed it to `ResentBlockSlot` :)

**Review by rkapka — APPROVED** [review]

**Pull request merged**